service/raft/raft_group_registry.cc

```diff
@@ -174,12 +174,28 @@ void raft_group_registry::init_rpc_verbs() {
         });
     });
 
-    ser::raft_rpc_verbs::register_raft_append_entries(&_ms, [handle_raft_rpc] (const rpc::client_info& cinfo, rpc::opt_time_point timeout,
-           raft::group_id gid, raft::server_id from, raft::server_id dst, raft::append_request append_request) mutable {
-        return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request)] (raft_rpc& rpc) mutable {
+    ser::raft_rpc_verbs::register_raft_append_entries(&_ms, [handle_raft_rpc, this] (const rpc::client_info& cinfo, rpc::opt_time_point timeout,
+        raft::group_id gid, raft::server_id from, raft::server_id dst, raft::append_request append_request) mutable {
+
+        // lw_shared_ptr (raft::log_entry_ptr) doesn't support cross-shard ref counting (see debug_shared_ptr_counter_type),
+        // so here we are copying entries to the target shard if it isn't equal the current one.
+        lw_shared_ptr<std::vector<raft::log_entry_ptr>> entries;
+        auto* entries_ptr = entries.get();
+        if (shard_for_group(gid) != this_shard_id()) {
+            entries = std::move(append_request.entries);
+            append_request.entries.clear();
+            entries_ptr = entries.get();
+        }
+        return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), entries_ptr] (raft_rpc& rpc) mutable {
+            if (entries_ptr) {
+                append_request.entries.reserve(entries_ptr->size());
+                for (auto& e: *entries_ptr) {
+                    append_request.entries.push_back(make_lw_shared(*e));
+                }
+            }
             rpc.append_entries(std::move(from), std::move(append_request));
             return netw::messaging_service::no_wait();
-        });
+        }).finally([entries = std::move(entries)](){});
```

Comment:
We are adding a continuation here even for the case where no copy is needed. I think we can do without this shared pointer. Something like that :
```
        return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), cpu = this_cpu()] (raft_rpc& rpc) mutable {
            if (this_cpu() != cpu) {
                raft::append_request r;
                r.current_term = append_request.current_term;
                r.prev...
                ....
                r.entries.reserve(r.entries.size());
                for (auto& e: r.entries) {
                    append_request.entries.push_back(make_lw_shared(*e));
                }
               rpc.append_entries(std::move(from), std::move(r));
            } else {
                rpc.append_entries(std::move(from), std::move(append_request));
            }
            return netw::messaging_service::no_wait();
    }
```

Reply:
Also for bisectability it is better to reorder last two patches.

Reply:
> We are adding a continuation here even for the case where no copy is needed. I think we can do without this shared pointer. Something like that :
> 
> ```
>         return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), cpu = this_cpu()] (raft_rpc& rpc) mutable {
>             if (this_cpu() != cpu) {
>                 raft::append_request r;
>                 r.current_term = append_request.current_term;
>                 r.prev...
>                 ....
>                 r.entries.reserve(r.entries.size());
>                 for (auto& e: r.entries) {
>                     append_request.entries.push_back(make_lw_shared(*e));
>                 }
>                rpc.append_entries(std::move(from), std::move(r));
>             } else {
>                 rpc.append_entries(std::move(from), std::move(append_request));
>             }
>             return netw::messaging_service::no_wait();
>     }
> ```

this would imply decrementing the source lw_shared_ptr's counter on the target core in the destructor -> crash

Reply:
> Also for bisectability it is better to reorder last two patches.

How specifically it is better?

Reply:
> > Also for bisectability it is better to reorder last two patches.
> 
> How specifically it is better?

It will be bisectable as opposite to not be bisectable in debug mode.

Reply:
> > We are adding a continuation here even for the case where no copy is needed. I think we can do without this shared pointer. Something like that :
> > ```
> >         return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), cpu = this_cpu()] (raft_rpc& rpc) mutable {
> >             if (this_cpu() != cpu) {
> >                 raft::append_request r;
> >                 r.current_term = append_request.current_term;
> >                 r.prev...
> >                 ....
> >                 r.entries.reserve(r.entries.size());
> >                 for (auto& e: r.entries) {
> >                     append_request.entries.push_back(make_lw_shared(*e));
> >                 }
> >                rpc.append_entries(std::move(from), std::move(r));
> >             } else {
> >                 rpc.append_entries(std::move(from), std::move(append_request));
> >             }
> >             return netw::messaging_service::no_wait();
> >     }
> > ```
> 
> this would imply decrementing the source lw_shared_ptr's counter on the target core in the destructor -> crash

Why? The lambda object, where original request is captured, is in the src shard, no?

Reply:
You are right, I missed that, fixed.

Reply:
> Also for bisectability it is better to reorder last two patches.

I don't understand, what bisectability means here.

Reply:
If each individual commit compiles and pass tests it is easier to do bisects if a bad commit does gets in (for instance the failure is rare). If the copy patch is after the registration patch there will be a test failure between them. If they are reordered there will not. Well, may be in this case this is not so important because we already have a failure, though different one...

Reply:
OK